src/dune_pkg/opam_repo.ml

```diff
@@ -132,7 +143,45 @@ let of_opam_repo_dir_path ~source ~repo_id opam_repo_dir_path =
   let serializable =
     Option.map source ~f:(fun source -> { Serializable.repo_id; source })
   in
-  { packages_dir_path; serializable }
+  { source = Directory packages_dir_path; serializable }
+;;
+
+let xdg_repo_location =
+  let ( / ) = Filename.concat in
+  lazy (Xdg.cache_dir (Lazy.force Dune_util.xdg) / "dune/git-repo" |> Path.of_string)
+;;
+
+let of_git_repo ~git ~repo_id source =
+  let open Fiber.O in
+  let dir = Lazy.force xdg_repo_location in
+  let* repo = Fetch.Git_repo.load_or_create ~git ~dir in
+  let* remote = Fetch.Git_repo.add_repo repo ~source in
+  let+ at_rev, computed_repo_id =
+    match repo_id with
+    | Some repo_id' ->
+      let+ at_rev = Fetch.Git_repo.Remote.rev_of_repository_id remote repo_id' in
+      at_rev, repo_id
+    | None ->
+      let* branch = Fetch.Git_repo.Remote.default_branch remote in
+      let name =
+        match branch with
+        | Some name -> name
+        | None ->
+          Code_error.raise
```

Comment:
If this is possible, it shouldn't be a `Code_error`. `User_error` would be more appropriate.